### ISSUE TITLE
Ask for obfs4

### ISSUE THREAD

**Issue by johncole2017** [n1nj4sec/pupy#678]
Hi, can i please ask you add obfs4 to transfering protocols ?
thanks

**Comment by Strazzom** [comment]
Scramblesuit has been outdated for a while now. The evaluation [here](https://trac.torproject.org/projects/tor/wiki/doc/PluggableTransports/ScrambleSuitEvaluation) indicates that it has been superseded by obfs4. I don't think scramblesuit is _broken_ per se, but it is not the standard for Tor bridges anymore because obfs4 is better.

There is an argument to be made for the importance of traffic obfuscation. Most of Pupy's transports stand out compared to regular network traffic (SSL excepted). In general, I think SSL is a better way to handle DPI, but it is nice to have the obfs4 alternative.

That being said, scramblesuit isn't broken, and implementing new features is time consuming. @alxchk can implement it if he has the time/inclination, but personally I don't consider it a high priority.

**Comment by alxchk** [comment]
https://www.maikel.pro/blog/progress-in-censorship-circumvention-tor-and-pluggable-transports-an-overview/

**Comment by Strazzom** [comment]
Interesting read.

> Obfs4 improves ScrambleSuit by providing the same security properties as ScrambleSuit but also includes an authenticated key exchange via BridgeDB. [...] The fact that obfs4 also provides authenticity which ScrambleSuit does not, also makes this protocol sound against active Man In the Middle Attacks by the Agent that controls the network.

Based on my reading of this, it sounds like the obfs4 modifications to scramblesuit were mainly that they added authentication via the elliptic curve public key. This was mainly to prevent active probing, where an adversary tries to connect to the bridge and see if it responds like a Tor node.

In theory, the lack of authentication in scramblesuit makes this easier to do. So, (for instance) if a target on an engagement wanted to know if the traffic leaving their network was using scramblesuit for obfuscation, they could potentially find out by actively probing the Pupy server, to see if it accepts the scramblesuit handshake. Granted, this is still difficult because of the time based element in this protocol, but it might be possible in theory.

The question becomes: is there a tool for fingerprinting scramblesuit with active probing, is it integrated into any enterprise network monitoring/security tools, and is it well known enough that blue team might actually deploy it?

Assuming there is a NIDS that is actually doing this by default, (probably not if it involves active probing, for legal reasons), then you are screwed. The likelihood of that being the case is slim, IMO.

Thoughts?

**Comment by alxchk** [comment]
> Thoughts?

My red team experience came from enterprise. And there things are very simple. Egress either explicitly blocked, or fully inspected (i.e. unknown blobs like any custom protocol will be blocked), or there is a hole )

I never saw anything more complex than that. Any unknown traffic either allowed (because of some kind of hole), or blocked with alerts. Period.

This obfs/scramblesuit things works only in environments where egress can not be blocked by default, but owner of the perimeter cares about the content. This likely the case for some countries, but not for small company or enterprise where you are likely working.. 

So, my thoughts are simple. I don't care about these transports ) They were before I touched this project, so I'm taking care they works. But no more than that.

**Comment by Strazzom** [comment]
If it is a simple protocol whitelist/blacklist, it might be worth looking into the FTE transport. It was rejected from Tor due to a lack of research and auditing, but in this case it isn't a life or death situation, and it is able to mimic many common protocols. They also claim it can "learn" to mimic uncommon protocols by analyzing packet samples.

Not sure on the details of any of this, but it could be worth looking into.

If obfs4 is off the table, I recommend closing this issue.

**Comment by johncole2017** [comment]
I must say this too, the Original one.
pupy uses ReflectiveDLLinject and this method detects by most of AVs ,
I want to know is there any way to bypass this or use another module for not detect by AVs ?

My first question goal is this : I want use pupy without any intruder, and not taken by any AVs .
thanks a lot

**Comment by alxchk** [comment]
This doesnt have relation to obfs4.

Don't use inject

**Comment by Strazzom** [comment]
@n1nj4sec Where is the source code for the audio transport? That is very similar to what BadBIOS was doing if I recall.

https://en.wikipedia.org/wiki/BadBIOS

**Comment by n1nj4sec** [comment]
Except my payload was emitting audible bips x). I can't find it I think it's on one of my raspberry SD card. Anyway it was just using some pyaudio recording and converted it to data
